src/models/geograph.py

```diff
@@ -614,21 +669,317 @@ def get_graph_components(
+            created with no edges. Defaults to None.
+            add_distance_edges (bool, optional): Boolean that determines whether
+            to add edges between every pair of nodes, with the distance between
+            their corresponding polygons as an edge attribute. Defaults to False.
+        """
+        # pylint: disable=super-init-not-called
+        self.has_df: bool = True
+        self.graph: nx.Graph = nx.Graph()
+        self.components_list: List[set] = components_list
+        if df is not None:
+            self.has_distance_edges: bool = add_distance_edges
+            self.df = self._load_from_dataframe(df, add_distance_edges)
+        else:
+            self.has_df = False
+            self.has_distance_edges = False
+            self.graph = nx.empty_graph(len(self.components_list))
+        self.metrics: Dict[str, Metric] = {}
+
+    def _load_from_dataframe(
+        self,
+        df: gpd.GeoDataFrame,
+        tolerance: float = 0.0,
+    ) -> gpd.GeoDataFrame:
+        """
+        Load graph from dataframe.
+
+        If `self.has_distance_edges` is True, then this will load a complete graph
+        and add the distance between each pair of polygons as an edge attribute.
+        Otherwise, it will load an empty graph.
+
+        In both cases, information about the polygons will be stored as node
+        attributes.
+
+        Args:
+            df (gpd.GeoDataFrame): GeoDataFrame containing polygon objects.
+            tolerance (float, optional): Redundant argument to ensure interface
+            is consistent with parent class. Defaults to 0.0.
+
+        Returns:
+            gpd.GeoDataFrame: A processed GeoDataFrame containing polygon objects.
+        """
+        # Reset index to ensure consistent indices
+        df = df.reset_index(drop=True)
+        # Using this list and iterating through it is slightly faster than
+        # iterating through df due to the dataframe overhead
+        geom: List[shapely.Polygon] = df["geometry"].tolist()
+        if self.has_distance_edges:
+            self.graph = nx.complete_graph(len(df))
+        else:
+            self.graph = nx.empty_graph(len(df))
+        # Add node attributes
+        for node in tqdm(
+            self.graph.nodes, desc="Constructing graph", total=len(self.graph)
+        ):
+            polygon = geom[node]
+            self.graph.nodes[node]["rep_point"] = polygon.representative_point()
+            self.graph.nodes[node]["area"] = polygon.area
+            self.graph.nodes[node]["perimeter"] = polygon.length
+            self.graph.nodes[node]["bounds"] = polygon.bounds
+
+        # Add edge attributes if necessary
+        if self.has_distance_edges:
+            for u, v, attrs in tqdm(
+                self.graph.edges.data(data=True),
+                desc="Calculating edge weights",
+                total=len(self.graph.edges),
+            ):
+                attrs["distance"] = geom[u].distance(geom[v])
```

Comment:
That loop and distance calculation is probably the very slow computation you mentioned right?

Reply:
Yeah that's that one